### DISCUSSION TITLE
Extra output from MaterialVectorPostprocessor in Transient simulation?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#20761] [Q&A General] [answered]
I'm trying to use a MaterialVectorPostprocessor to view the values of my material property at quadrature points. When I do this in a transient simulation, it seems I'm getting the full history in each csv file up to the current time. The current [MaterialVectorPostprocessor test](https://github.com/idaholab/moose/blob/next/test/tests/vectorpostprocessors/material_vector_postprocessor/basic.i) is a steady state test, so this behavior isn't seen. But if I modify the input file to do a Transient simulation like this:

```
# test that all scalar material properties are properly recorded in basic usage.
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 10
  ny = 10
[]

[Variables]
  [./u]
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]
  [./time]
    type = TimeDerivative
    variable = u
  [../]
[]

[BCs]
  [./left]
    type = DirichletBC
    variable = u
    boundary = left
    value = 0
  [../]
  [./right]
    type = DirichletBC
    variable = u
    boundary = right
    value = 1
  [../]
[]

[Materials]
  [./mat]
    type = GenericConstantMaterial
    prop_names = 'prop1 prop2 prop3'
    prop_values = '1 2 42'
  [../]
[]

[VectorPostprocessors]
  [./vpp]
    type = MaterialVectorPostprocessor
    material = 'mat'
    elem_ids = '3 4 7 42 88'
  [../]
[]

[Executioner]
  type = Transient
  dt = 1
  num_steps = 3
  solve_type = PJFNK
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]

[Outputs]
  execute_on = 'initial timestep_end'
  csv = true
[]
```
then each CSV file is longer than the previous and presumably contains data from previous timesteps. This does not seem like the correct behavior as other Transient VPP tests will output more or less the same size number of lines on each timestep.

**Comment by GiudGiud** [comment, marked as answer]
what do you get with this flag `contains_complete_history` ?
Do you get a single file with the complete history or do you get another dissimilar behavior?

What you describe seems to be in between what this option should do and what should be done normally. 

If we can confirm this we ll raise an issue, thanks for reporting

**Reply by jessecarterMOOSE** [reply]
When I set `contains_complete_history = false`, I see multiple csv files rather than one big one, but each one seems to contain information from all timesteps before it. If I set that flag to true, then I'm not seeing what I expect either. There is a single csv file, but, for instance, info from element 88 seems to be printed 6 times but element 3 only once? Doesn't seem right.

**Reply by jessecarterMOOSE** [reply]
@GiudGiud FWIW I don't see a call to `initialize()` in [MaterialVectorPostprocessor.C](https://github.com/idaholab/moose/blob/next/framework/src/vectorpostprocessors/MaterialVectorPostprocessor.C), so is it possible we are not resetting the vectors after each iteration?

**Reply by GiudGiud** [reply]
Might just be that. I have not gotten to verifying this

**Reply by jessecarterMOOSE** [reply]
I tried adding this:
```
void
MaterialVectorPostprocessor::initialize()
{
    _elem_ids.clear();
    _qp_ids.clear();
}
```
Which sort of works - the values are no longer duplicated in the output, but a bunch of zeros also get written out, and more of them when I run in parallel.

**Reply by jessecarterMOOSE** [reply]
No, that's not it. The material properties are static in this simulation and they are still being carried from one timestep to the next.

In the meantime, is [Sampler1DReal](https://mooseframework.inl.gov/modules/thermal_hydraulics/source/vectorpostprocessors/Sampler1DReal.html) a good substitute? Only differences as far as I can tell is you specify blocks rather than element numbers. You don't get the element and qp number either, but that's ok.

**Reply by GiudGiud** [reply]
only if you are using the thermal hydraulics module

**Reply by GiudGiud** [reply]
well that might be an exageration. But you will definitely need to link the module in your application to use it

**Reply by jessecarterMOOSE** [reply]
I'm working in a Stork app so I can just edit the Makefile to bring it in. I would assume it's only a temporary solution until someone can fix the VPP.

**Reply by GiudGiud** [reply]
I can reproduce this.
I think it's just the code that uses push_back everywhere, adding in the vectors instead of replacing the values. 
I ll make a PR

**Reply by jessecarterMOOSE** [reply]
Great, thanks!
